News UI overhaul and PDA notifications

juliangiebel opened the pull request:
## About the PR
This PR improves the news writer UI, cleans up `NewsSystem.cs` somewhat and adds PDA notifications which can be used for other things in the future (PDA messaging).

## Open questions
~~Should I add a `PDA` or `Notifications` channel?~~
I added a `Notifications` channel

## Why / Balance
News code and the news writer UI needed some improvement.
`NewsSystem.cs` called the PDA ringer directly so I took the opportunity to implement generic PDA notifications for cartridges.

This PR also adds a new style class for buttons that uses its own background texture. The `ButtonSmall`:
![image](https://github.com/space-wizards/space-station-14/assets/8915246/a72a746b-16e6-4724-afc1-44aa34083d57)


## Technical details
Mostly just changed UI and code refactoring. PDA notifications are sent by calling a method on `CartridgeLoaderSystem.cs` that checks if the cartridge loader (PDA) is muted and if it isn't raises an event on itself for actually sending the notification. This is done to keep cartridges decoupled from PDA code.

## Media
Empty mass media console:
![image](https://github.com/space-wizards/space-station-14/assets/8915246/4603ad65-95ae-4e0c-977b-6ddc5034bee4)

Article editor:
![image](https://github.com/space-wizards/space-station-14/assets/8915246/c3091494-cf81-4501-a9d2-ffba2b167854)
*(I corrected the heading size being stated as allowing 1-5 to 1-3 already)*

Article preview:
![image](https://github.com/space-wizards/space-station-14/assets/8915246/85bce557-d154-40ec-8ecb-28cc4fb0791c)

News article list with small button and visible button timer:
![image](https://github.com/space-wizards/space-station-14/assets/8915246/197e88b4-9c54-4f71-9c2f-111cf6a502a0)

PDA notifications:
![image](https://github.com/space-wizards/space-station-14/assets/8915246/a2ca589c-51d1-4ccc-b49e-0728357efd41)

- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes
<!--
List any breaking changes, including namespace, public class/method/field changes, prototype renames; and provide instructions for fixing them. This will be pasted in #codebase-changes.
-->
News code namespaces changes. ...Write... and ...Read... got renamed into ...Writer... and ...Reader... . Misc typo fixes for method and class names.

**Changelog**
<!--
Make players aware of new features and changes that could affect how they play the game by adding a Changelog entry. Please read the Changelog guidelines located at: https://docs.spacestation14.io/en/getting-started/pr-guideline#changelog
-->

:cl:
- tweak: The mass media consoles UI got overhauled
- add: Added PDA notifications in chat for new news articles. They can be turned off in the news reader program and are only visible to you.

Requires https://github.com/space-wizards/RobustToolbox/pull/4332
Closes #23968

AJCM-git commented on Content.Client/MassMedia/Ui/ArticleEditorPanel.xaml.cs:
```diff
@@ -0,0 +1,93 @@
+        ButtonPublish.StyleClasses.Add(StyleBase.ButtonOpenLeft);
+        ButtonPublish.StyleClasses.Add(StyleNano.StyleClassButtonColorGreen);
+
+        ContentField.GetChild(0).Margin = new Thickness(9, 3);
+        var scrollbar = ContentField.GetChild(1);
+        scrollbar.SetWidth = 6f;
+        scrollbar.Margin = new Thickness(9, 0, 2 , 0);
```
is there a way to move margins value to the xaml to avoid hardcoding?

juliangiebel replied:
duplicate

AJCM-git commented on Content.Client/MassMedia/Ui/NewsWriterMenu.xaml.cs:
```diff
@@ -0,0 +1,96 @@
+
+        ContentsContainer.RectClipContent = false;
+        var scrollbar = ArticleListScrollbar.GetChild(1);
+        scrollbar.SetWidth = 6f;
+        scrollbar.Margin = new Thickness(0, 0, 2 , 0);
```
a

AJCM-git commented on Content.Client/MassMedia/Ui/NewsWriterMenu.xaml.cs:
```diff
@@ -0,0 +1,96 @@
+                Author = article.Author ?? Loc.GetString("news-read-ui-no-author"),
+                PublicationTime = article.ShareTime,
+                ArtcileNumber = i,
+                Margin = new Thickness(0, 0, 0, 8)
```
Why not make this the default thickness of the NewsArticleCard?

AJCM-git commented on Content.Client/MassMedia/Ui/NewsWriterMenu.xaml.cs:
```diff
@@ -0,0 +1,96 @@
+    {
+        base.Resized();
+        var margin = ArticleEditorPanel.Margin;
+        ArticleEditorPanel.Margin =  new Thickness(Width - 1, margin.Top, margin.Right, margin.Bottom);
```
this looks like a bandaid to a bug, if so this is fine but leave a comment

AJCM-git commented on Content.Server/PDA/PdaSystem.cs:
```diff
@@ -102,6 +108,27 @@ private void UpdateAllPdaUisOnStation()
+                return;
+
+            var message = FormattedMessage.EscapeText(args.Message);
+            var wrappedMessage = $"[font size=12][bold]PDA[/bold] {args.Header}: [/font]\n\"{message}\"";
```
localization (make this a ftl string, they support markup)

AJCM-git reviewed (changes requested):
fix build

juliangiebel commented on Content.Client/MassMedia/Ui/NewsWriterBoundUserInterface.cs:
```diff
@@ -0,0 +1,89 @@
+        [ViewVariables]
+        private NewsWriterMenu? _menu;
+
+        [Dependency] private readonly IEntitySystemManager _entitySystem = default!;
```
Looks like I missed that. It shows as me adding that line because I renamed the file. Sorry if that made reviewing more annoying 🤦‍♂️

juliangiebel commented on Content.Client/MassMedia/Ui/ArticleEditorPanel.xaml.cs:
```diff
@@ -0,0 +1,93 @@
+        ButtonPublish.StyleClasses.Add(StyleBase.ButtonOpenLeft);
+        ButtonPublish.StyleClasses.Add(StyleNano.StyleClassButtonColorGreen);
+
+        ContentField.GetChild(0).Margin = new Thickness(9, 3);
+        var scrollbar = ContentField.GetChild(1);
+        scrollbar.SetWidth = 6f;
+        scrollbar.Margin = new Thickness(9, 0, 2 , 0);
```
Not for the scrollbar, as you can only set new children using xaml and not access already present ones afaik. The bar width and margin should be a property on the scroll container tbh but that should be a separate engine PR

AJCM-git commented on Content.Client/MassMedia/Ui/NewsWriterBoundUserInterface.cs:
```diff
@@ -0,0 +1,87 @@
+
+        var stringContent = Rope.Collapse(_menu.ArticleEditorPanel.ContentField.TextRope);
+
+        if (stringContent.Length == 0 || _gameTicker == null)
+            return;
+
+        var name = title.Length <= 100 ? title.Trim() : $"{title.Trim()[..100]}...";
```
Uhhh shouldnt this be getting checked by the server anyways? (and if it isnt it really should)

juliangiebel replied:
It does, this is here to spare the networking code from handeling massive copy pasta (which is an issue even when the entity system truncates the string)

AJCM-git replied:
aight

AJCM-git commented on Content.Client/MassMedia/Ui/NewsWriterMenu.xaml.cs:
```diff
@@ -0,0 +1,97 @@
+[GenerateTypedNameReferences]
+public sealed partial class NewsWriterMenu : FancyWindow
+{
+    private readonly IGameTiming _gameTiming;
```
Use dependency attribute and IoCManager.InjectDependencies(this)

juliangiebel replied:
Aren't we supposed to limit the use of IoCManagert.InjectDependencies?

AJCM-git replied:
i dont think so? also all windows follow the pattern with InjectDependencies

AJCM-git reviewed (changes requested):
Dont really have much time for a deep code dive, so take a look if you need to update anything to `NetEntity` stuff, but i can do some mechanic feedback
- The next and past (should be renamed to prev for consistency, actually should prob just use `▶` and `◀` and have tooltips) buttons shouldnt be shown if there are no pages, also they should gray out when there are no more articles forward or backwards.
- The silence news notifs button needs a tooltip
- Clicking on a "New article" page title does not move the window (only when you click the Mass media management one)
- The "new article" preview lets you still edit the title
- I know it wasnt your fault but can you rename the mass-media console to news management console or something (or at least get rid of the - and call it mass media for consistency)?
- A confirmation for article deletion would be nice (i think admin menu does this if you need an example)
- Maaaaaybe add a StationNewsComponent to any station a NewsWriter spawns on (maybe a bad idea? idk)?

juliangiebel commented:
> Clicking on a "New article" page title does not move the window (only when you click the Mass media management one)

I don't know how to fix that one tbh.

> Maaaaaybe add a StationNewsComponent to any station a NewsWriter spawns on (maybe a bad idea? idk)?

Doesn't every station come with a `StationNewsComponent` if it's in general station prototype?

---
Thanks for the feedback btw.

AJCM-git commented:
> > Clicking on a "New article" page title does not move the window (only when you click the Mass media management one)
> 
> I don't know how to fix that one tbh.

yeah, i think it would need an engine PR, disregard if you dont find a way

> > Maaaaaybe add a StationNewsComponent to any station a NewsWriter spawns on (maybe a bad idea? idk)?
> 
> Doesn't every station come with a `StationNewsComponent` if it's in the general station prototype?
> 

Indeed they do, but i dont see why they should have it if they dont have a writer (maybe some mapper will make one with default news in the future, but in that case they would have to add stuff manually anyways) and making it tied to the writer means you will be able to make station news IC from scratch without debug tools

AJCM-git commented:
@juliangiebel you coming back to this?

juliangiebel commented:
Yes. I've been ill the last 4 weeks and I won't be at home this weekend.
I'll try to get the reviews addressed next week.

metalgearsloth commented:
Re-open if you come back to it.

juliangiebel commented:
I addressed the remaining bit of reviews and implemented fancy confirm buttons. The `ConfirmButton` control is generic and can be just used instead of a button. It has dedicated style subclasses for its confirmation stage, changes the button text to a confirmation prompt when clicked once and invokes the `OnPressed` action when clicked twice.
The button reverts to normal when not pressed a second time after two second, which is a property on the control.
It also has a small delay before it accepts a second press to prevent accidentally confirming it.

https://github.com/space-wizards/space-station-14/assets/8915246/168c4e81-c0f1-4b14-8283-e53c70200b2a

AJCM-git commented on Content.Client/MassMedia/Ui/ArticleEditorPanel.xaml.cs:
```diff
@@ -0,0 +1,91 @@
+    {
+        RobustXamlLoader.Load(this);
+
+        ButtonPublish.StyleClasses.Add(StyleBase.ButtonOpenLeft);
+        ButtonPublish.StyleClasses.Add(StyleNano.StyleClassButtonColorGreen);
```
I think StyleClasses can be in the xaml? ButtonPublish.

juliangiebel replied:
I don't quite remember anymore but I tried setting them in xaml first but there was a problem with that.

AJCM-git reviewed (changes requested):
Just pushed some minor cleanup too annoying to leave reviews for, tried to separate the most important bits into commits so you can know what changed and see if you agree with it. Really good progress, all that's left to do are some minor things like:
- Does MultilineTextSubmit? I may have misunderstood how it's supposed to work cuz i couldnt get it to work
- Use new Subs.BUiEven pattern for bui messages handlers
- Is it really necessary to touch Content.Server.csproj for this PR? It's a minor and harmless change but it's the type of thing that i think needs to be separated
- I added a Vseparator between Author/Publish time that you may like to tweak slightly (make it thicker or add more padding) since it looks a little odd 
- Test saving maps with articles or with the console and see if something blows up
- There should prob be some clientside visual feedback for title and content limits, something simple like making the text or box red will do
- I dont think the pda news UI updates when you publish something if there are no articles (works if there are already one)
- There is client-side validation in both the bui and a window, are both of these necessary?

This is something thats probably better for another PR but the UI and some of the logic is REALLY tied to cartridges which is a shame since it makes adding things like the newsletter harder, it would be nice if it worked more like the crew manifest cartridge or something. Kinda out of scope so if you dont end up doing it make an issue for it.

juliangiebel commented on Content.Client/MassMedia/Ui/NewsWriterMenu.xaml:
Setting `RectClipContent` like this doesn't work. Idk. if there is a way to do that in xaml

AJCM-git replied:
Huh, thats very weird, revert that if you want then

juliangiebel commented:
It works. Only paper uses it.
Before I added that keybind paper would submit by just pressing enter, which kinda sucked.

juliangiebel commented:
> Test saving maps with articles or with the console and see if something blows up

Saving maps just breaks because of other issues

juliangiebel commented:
To restate what I said in discord here so it doesn't get lost:
> - There should prob be some clientside visual feedback for title and content limits, something simple like making the text or box red will do
> - I dont think the pda news UI updates when you publish something if there are no articles (works if there are already one)
> - There is client-side validation in both the bui and a window, are both of these necessary?

- I think that'd be the first time someone does that with our UI but I'll try implementing that.
- I tested it and it seemed to update for me
- I didn't see any duplicate client side validation. Can you tell me where it's at? Maybe I just missed it.

juliangiebel commented:
@AJCM-git I implemented visual feedback.
